File: engine/network/graph.py
"""The road network graph: nodes + directed edges + adjacency indexes."""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass, field

from engine.network.edge import Road
from engine.network.node import Node, NodeType
from engine.traffic.traffic_model import congestion_factor


@dataclass
class RoadNetwork:
    """Holds every node and road plus fast lookup indexes.

    ``adjacency[node_id]`` -> outgoing edge ids
    ``incoming[node_id]``  -> incoming edge ids
    """

    nodes: dict[str, Node] = field(default_factory=dict)
    edges: dict[str, Road] = field(default_factory=dict)
    adjacency: dict[str, list[str]] = field(default_factory=dict)
    incoming: dict[str, list[str]] = field(default_factory=dict)

    # ------------------------------------------------------------------ #
    # Builders
    # ------------------------------------------------------------------ #
    def add_node(self, node: Node) -> None:
        if node.id in self.nodes:
            raise ValueError(f"node {node.id} already exists")
        self.nodes[node.id] = node
        self.adjacency.setdefault(node.id, [])
        self.incoming.setdefault(node.id, [])

    def add_road(self, road: Road) -> None:
        if road.id in self.edges:
            raise ValueError(f"road {road.id} already exists")
        if road.source not in self.nodes or road.destination not in self.nodes:
            raise ValueError(
                f"road {road.id}: source/destination nodes must exist"
            )
        self.edges[road.id] = road
        self.adjacency[road.source].append(road.id)
        self.incoming[road.destination].append(road.id)

    # ------------------------------------------------------------------ #
    # Lookups
    # ------------------------------------------------------------------ #
    def node(self, node_id: str) -> Node:
        return self.nodes[node_id]

    def edge(self, edge_id: str) -> Road:
        return self.edges[edge_id]

    def node_count(self) -> int:
        return len(self.nodes)

    def edge_count(self) -> int:
        return len(self.edges)

    def outgoing_edges(self, node_id: str) -> list[str]:
        return self.adjacency.get(node_id, [])

    def incoming_edges(self, node_id: str) -> list[str]:
        return self.incoming.get(node_id, [])

    def neighbors(self, node_id: str) -> Iterable[tuple[str, str]]:
        """Yield ``(neighbor_id, edge_id)`` pairs for traversable roads."""
        for edge_id in self.adjacency.get(node_id, []):
            road = self.edges[edge_id]
            if road.is_traversable:
                yield road.destination, edge_id

    def edge_between(self, source: str, destination: str) -> Road | None:
        for edge_id in self.adjacency.get(source, []):
            if self.edges[edge_id].destination == destination:
                return self.edges[edge_id]
        return None

    def nodes_of_type(self, node_type: NodeType) -> list[Node]:
        return [n for n in self.nodes.values() if n.type is node_type]

    def road_summary(self) -> dict[str, object]:
        return {
            "nodes": self.node_count(),
            "edges": self.edge_count(),
            "closed": sum(
                1 for r in self.edges.values() if not r.is_traversable
            ),
        }

    def refresh_dynamic_state(self) -> None:
        """Recompute occupancy, congestion and dynamic travel times.

        Occupancy is derived from vehicles currently on each road; a single
        vehicle may occupy a road even while it is "between" it during a tick,
        so this is applied once per tick by the engine.
        """
        for road in self.edges.values():
            road.congestion = self._congestion(road)
            road.current_travel_time = road.base_travel_time * congestion_factor(
                road.congestion
            )

    @staticmethod
    def _congestion(road: Road) -> float:
        if road.capacity <= 0:
            return 1.0
        return road.current_vehicle_count / road.capacity

File: src/core/network/graph.ts
import { RoadNetwork, Node, Road } from '../../types';
import { updateRoadDynamicState, isTraversable } from './edge';

export function createRoadNetwork(): RoadNetwork {
  return {
    nodes: new Map(),
    edges: new Map(),
    adjacency: new Map(),
    incoming: new Map(),
  };
}

export function addNode(network: RoadNetwork, node: Node): void {
  if (network.nodes.has(node.id)) {
    throw new Error(`node ${node.id} already exists`);
  }
  network.nodes.set(node.id, node);
  network.adjacency.set(node.id, []);
  network.incoming.set(node.id, []);
}

export function addRoad(network: RoadNetwork, road: Road): void {
  if (network.edges.has(road.id)) {
    throw new Error(`road ${road.id} already exists`);
  }
  if (!network.nodes.has(road.source) || !network.nodes.has(road.destination)) {
    throw new Error(`road ${road.id}: source/destination nodes must exist`);
  }
  network.edges.set(road.id, road);
  network.adjacency.get(road.source)!.push(road.id);
  network.incoming.get(road.destination)!.push(road.id);
}

export function getNode(network: RoadNetwork, nodeId: string): Node {
  const node = network.nodes.get(nodeId);
  if (!node) throw new Error(`node ${nodeId} not found`);
  return node;
}

export function getRoad(network: RoadNetwork, edgeId: string): Road {
  const road = network.edges.get(edgeId);
  if (!road) throw new Error(`road ${edgeId} not found`);
  return road;
}

export function nodeCount(network: RoadNetwork): number {
  return network.nodes.size;
}

export function edgeCount(network: RoadNetwork): number {
  return network.edges.size;
}

export function outgoingEdges(network: RoadNetwork, nodeId: string): string[] {
  return network.adjacency.get(nodeId) ?? [];
}

export function incomingEdges(network: RoadNetwork, nodeId: string): string[] {
  return network.incoming.get(nodeId) ?? [];
}

export function* neighbors(network: RoadNetwork, nodeId: string): Generator<[string, string]> {
  for (const edgeId of network.adjacency.get(nodeId) ?? []) {
    const road = network.edges.get(edgeId)!;
    if (isTraversable(road)) {
      yield [road.destination, edgeId];
    }
  }
}

export function edgeBetween(network: RoadNetwork, source: string, destination: string): Road | null {
  for (const edgeId of network.adjacency.get(source) ?? []) {
    const road = network.edges.get(edgeId)!;
    if (road.destination === destination) return road;
  }
  return null;
}

export function nodesOfType(network: RoadNetwork, type: Node['type']): Node[] {
  const result: Node[] = [];
  for (const node of network.nodes.values()) {
    if (node.type === type) result.push(node);
  }
  return result;
}

export function roadSummary(network: RoadNetwork): { nodes: number; edges: number; closed: number } {
  let closed = 0;
  for (const road of network.edges.values()) {
    if (!isTraversable(road)) closed++;
  }
  return { nodes: network.nodes.size, edges: network.edges.size, closed };
}

export function refreshDynamicState(network: RoadNetwork, vehicleCounts: Map<string, number>): void {
  for (const [roadId, road] of network.edges) {
    const count = vehicleCounts.get(roadId) ?? 0;
    updateRoadDynamicState(road, count);
  }
}

export interface NetworkValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

export function validateNetwork(network: RoadNetwork): NetworkValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (network.nodes.size === 0) {
    errors.push('Network must contain at least one node');
  }
  if (network.edges.size === 0) {
    errors.push('Network must contain at least one edge');
  }

  // Check each node
  for (const [nodeId, node] of network.nodes) {
    if (!node.id || typeof node.x !== 'number' || typeof node.y !== 'number') {
      errors.push(`Invalid node definition for node ${nodeId}`);
    }
    const outEdges = network.adjacency.get(nodeId) ?? [];
    const inEdges = network.incoming.get(nodeId) ?? [];
    if (outEdges.length === 0 && inEdges.length === 0) {
      warnings.push(`Node ${nodeId} is isolated (no incoming or outgoing edges)`);
    }
  }

  // Check each edge
  for (const [edgeId, road] of network.edges) {
    if (!network.nodes.has(road.source)) {
      errors.push(`Road ${edgeId} references non-existent source node: ${road.source}`);
    }
    if (!network.nodes.has(road.destination)) {
      errors.push(`Road ${edgeId} references non-existent destination node: ${road.destination}`);
    }
    if (road.distance <= 0) {
      errors.push(`Road ${edgeId} has invalid distance: ${road.distance} (must be > 0)`);
    }
    if (road.speedLimit <= 0) {
      errors.push(`Road ${edgeId} has invalid speedLimit: ${road.speedLimit} (must be > 0)`);
    }
    if (road.capacity <= 0) {
      errors.push(`Road ${edgeId} has invalid capacity: ${road.capacity} (must be > 0)`);
    }
    if (road.lanes < 1) {
      errors.push(`Road ${edgeId} has invalid lane count: ${road.lanes} (must be >= 1)`);
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
  };
}